scvi/external/sysvi/_base_components.py

```diff
@@ -0,0 +1,345 @@
+        cov
+            tensor of covariate values with shape ``(n_cov,)`` or None
+
+        Returns
+        -------
+        py:class:`torch.Tensor`
+            tensor of shape ``(n_out,)``
+
+        """
+        for i, layers in enumerate(self.fc_layers):
+            for layer in layers:
+                if layer is not None:
+                    if isinstance(layer, BatchNorm1d):
+                        if x.dim() == 3:
+                            x = torch.cat([(layer(slice_x)).unsqueeze(0) for slice_x in x], dim=0)
+                        else:
+                            x = layer(x)
+                    else:
+                        # Injection of covariates
+                        if (
+                            self.n_cov > 0
+                            and isinstance(layer, Linear)
+                            and self.inject_into_layer(i)
+                        ):
+                            x = torch.cat((x, cov), dim=-1)
+                        x = layer(x)
+        return x
+
+
+class VarEncoder(Module):
+    """Encode variance (strictly positive).
+
+    Parameters
+    ----------
+    n_input
+        Number of input dimensions, used if mode is sample_feature
+    n_output
+        Number of variances to predict
+    mode
+        How to compute var
+        'sample_feature' - learn per sample and feature
+        'feature' - learn per feature, constant across samples
+        'linear' - linear with respect to input mean, var = a1 * mean + a0;
+         not suggested to be used due to bad implementation for positive constraining
+    eps
+    """
+
+    def __init__(
+        self,
+        n_input: int,
+        n_output: int,
+        mode: Literal["sample_feature", "feature", "linear"],
+        eps: float = 1e-4,
+    ):
+        super().__init__()
+
+        self.eps = eps
+        self.mode = mode
+        if self.mode == "sample_feature":
+            self.encoder = Linear(n_input, n_output)
+        elif self.mode == "feature":
+            self.var_param = Parameter(torch.zeros(1, n_output))
+        elif self.mode == "linear":
+            self.var_param_a1 = Parameter(torch.tensor([1.0]))
+            self.var_param_a0 = Parameter(torch.tensor([self.eps]))
+        else:
+            raise ValueError("Mode not recognised.")
+        self.activation = torch.exp
+
+    def forward(self, x: torch.Tensor, x_m: torch.Tensor):
+        """Forward pass through model
+
+        Parameters
+        ----------
+        x
+            Used to encode var if mode is sample_feature; dim = n_samples x n_input
+        x_m
+            Used to predict var instead of x if mode is linear; dim = n_samples x 1
+
+        Returns
+        -------
+        Predicted var
+        """
+        # Force to be non nan - TODO come up with better way to do so
+        if self.mode == "sample_feature":
+            v = self.encoder(x)
+            v = (
+                torch.nan_to_num(self.activation(v)) + self.eps
```

Comment:
Please try to avoid nan_to_num

Reply:
what would you suggest?

Reply:
Using softplus activation will be much more stable than exp. Do you need exp here for specific reasons? Clamping v would otherwise be safe (something like 20).